waspc/tests/Generator/Valid/PackageJsonTest.hs

```diff
@@ -0,0 +1,183 @@
+        let fieldName = depTypeToFieldName depType
+
+        testFailure
+          (makeRequiredDepValidator depType ("required-pkg", "1.0.0"))
+          (pkgJson `withDep` (oppositeDepType, ("required-pkg", "1.0.0")))
+          ["Wasp requires package \"required-pkg\" to be in \"" <> fieldName <> "\"."]
+
+    it "sets correct field path" $
+      forEachCase $ \depType pkgJson -> do
+        let fieldName = depTypeToFieldName depType
+
+        testFieldPath
+          (makeRequiredDepValidator depType ("required-pkg", "1.0.0"))
+          (pkgJson `withDep` (depType, ("required-pkg", "2.0.0")))
+          [fieldName, "required-pkg"]
+
+  describe "inDependency" $ do
+    it "runs validator on dependency value" $
+      forEachCase $ \depType pkgJson -> do
+        let isExpectedValidator = V.eqJust "1.0.0"
+
+        testSuccess
+          (inDependency depType "test-pkg" isExpectedValidator)
+          (pkgJson `withDep` (depType, ("test-pkg", "1.0.0")))
+
+    it "passes Nothing when dependency is missing" $
+      forEachCase $ \depType pkgJson -> do
+        let isNothingValidator Nothing = V.success
+            isNothingValidator _ = V.failure "Should be Nothing"
+
+        testSuccess
+          (inDependency depType "missing-pkg" isNothingValidator)
+          pkgJson
+
+    it "sets correct field path with nested fields" $
+      forEachCase $ \depType pkgJson -> do
+        let fieldName = depTypeToFieldName depType
+
+        let alwaysFailValidator = const $ V.failure "test error"
+        let validator =
+              inDependency depType "test-pkg" alwaysFailValidator
+
+        testFieldPath
+          validator
+          pkgJson
+          [fieldName, "test-pkg"]
+
+    it "looks in correct dependency type only" $ do
+      let pkgJson =
+            mockPackageJsonEmpty
+              `withDep` (Runtime, ("pkg-a", "1.0.0"))
+              `withDep` (Development, ("pkg-b", "2.0.0"))
+
+      testSuccess
+        (inDependency Runtime "pkg-a" $ V.eqJust "1.0.0")
+        pkgJson
+
+      testSuccess
+        (inDependency Development "pkg-b" $ V.eqJust "2.0.0")
+        pkgJson
+
+-- | Runs a test for each combination of DependencyType and empty/mock PackageJson
+forEachCase :: (DependencyType -> P.PackageJson -> Expectation) -> Expectation
+forEachCase fn =
+  sequence_
+    [ fn depType pkgJson
+      | depType <- [Runtime, Development],
+        pkgJson <- [mockPackageJson, mockPackageJsonEmpty]
+    ]
+
+testSuccess :: V.Validator P.PackageJson -> P.PackageJson -> IO ()
+testSuccess validator pkgJson =
+  V.execValidator validator pkgJson `shouldBe` []
+
+testFailure :: V.Validator P.PackageJson -> P.PackageJson -> [String] -> IO ()
+testFailure validator pkgJson expectedErrors = do
+  let errors = V.execValidator validator pkgJson
+  (V.message <$> errors) `shouldBe` expectedErrors
+
+testFieldPath :: V.Validator P.PackageJson -> P.PackageJson -> [String] -> IO ()
```

Comment:
Took a bit of time to fgiure out what this function is doing.
So it is what, checking at each of errors has correct field path in it?
Maybe let's name it `testErrorsHaveCorrectFieldPath`?